### DISCUSSION TITLE
WinSSHTerm - does it work with Teleport?

### DISCUSSION THREAD

**Discussion by WinSSHTerm** [gravitational/teleport#45040] [Q&A] [answered]
Hi,
a user [recently asked](https://www.reddit.com/r/WinSSHTerm/comments/1e68s57/ssh_connection_to_teleport/) if it is possible to use [WinSSHTerm](https://winsshterm.blogspot.com/) (which is basically a wrapper GUI for PuTTY and WinSCP) to ssh into a server through Teleport. WinSSHTerm now supports certificates (since [2.37.1](https://github.com/WinSSHTerm/WinSSHTerm/releases/tag/2.37.1)), however I don't have access to a Teleport setup.

Here is a [YouTube video](https://www.youtube.com/watch?v=Cjv5aqkUh2M) that shows how to manually set up certificates to launch PuTTY and WinSCP. For use with Teleport, a proxy ("Local") has to be configured too afaik.

There is a [migration tool](https://github.com/P-St/Migrate2WinSSHTerm) available, which should automatically create a connection file for WinSSHTerm and migrate all relevant settings from the PuTTY sessions generated by Teleport.

I'd be glad if someone could check it out and confirm if it works, thanks!
Patrick

**Comment by webvictim** [comment]
Hey - I tried to run the migration tool and got an error: https://github.com/P-St/Migrate2WinSSHTerm/issues/13

I'll be happy to test things if it's possible to fix that error. Thanks!

Edit: now fixed!

**Comment by webvictim** [comment]
OK, now I've got the migration tool working, I ran it and it generated a `connections.xml` file containing the details of my two PuTTY connections that were added using [Teleport's `tsh puttyconfig`](https://goteleport.com/docs/connect-your-client/putty-winscp/#summary) command.

The connections load OK, but when I attempt to load a session to connect through Teleport, I get an error: "X11 forwarding is not enabled":

![Screenshot 2024-08-08 at 16 35 40](https://github.com/user-attachments/assets/c2fd071d-4ce1-495c-8df2-d1376152f054)

The session halts at this point and I cannot access the server.

This error is being passed back from Teleport, which does not allow X forwarding by default. As such, the PuTTY sessions it creates don't have X forwarding enabled when created either. It seems that out of the box, WinSSHTerm expects this to be enabled by default?

If I manually edit the "Config" of the connection to set "don't forward" next to "X11 forward", things work as expected:

![image](https://github.com/user-attachments/assets/e3f8b89d-d69e-4a32-b2ed-e2edf801f239)

### Debugging info

`connections.xml` (real hostname was redacted):
```xml
<?xml version='1.0' encoding='utf-8'?>
<WinSSHTerm Version='1'>
<Node Name='Sessions' Type='Container' Expanded='True'>
<Node Name='ip-172-31-30-140%20(proxy:teleport.example.com)' Type='Connection' Descr='' Username='ubuntu' Password='' PrivateKey='C:\Users\gus\.tsh\keys\teleport.example.com\webvictim.ppk' Hostname='ip-172-31-30-140' Port='3022' Certificate='C:\Users\gus\.tsh\keys\teleport.example.com\webvictim-ssh\purple-cert.pub' pSshProxy='enabled' pType='Local' pHost='teleport.example.com' pPort='0' pUser='ubuntu' pTelnetCmd='C:\\Windows\\tsh.exe proxy ssh --cluster=purple --proxy=%proxyhost %user@%host:%port' />
<Node Name='ip-172-31-8-63%20(proxy:teleport.example.com)' Type='Connection' Descr='' Username='ubuntu' Password='' PrivateKey='C:\Users\gus\.tsh\keys\teleport.example.com\webvictim.ppk' Hostname='ip-172-31-8-63' Port='22' Certificate='C:\Users\gus\.tsh\keys\teleport.example.com\webvictim-ssh\purple-cert.pub' pSshProxy='enabled' pType='Local' pHost='teleport.example.com' pPort='0' pUser='ubuntu' pTelnetCmd='C:\\Windows\\tsh.exe proxy ssh --cluster=purple --proxy=%proxyhost %user@%host:%port' />
</Node>
</WinSSHTerm>
```

Screenshot of the `Computer\HKEY_CURRENT_USER\Software\SimonTatham\PuTTY\Sessions\ip-172-31-30-140%20(proxy:teleport.example.com)` registry key as created by Teleport with `tsh puttyconfig`:

![Screenshot 2024-08-08 at 16 40 32](https://github.com/user-attachments/assets/f9268562-f49b-4666-9337-0798fe3920f0)

**Comment by webvictim** [comment]
I also tried to use the "Copy Files" button on the connection to invoke WinSCP, but unfortunately that doesn't work:

![Screenshot 2024-08-08 at 16 46 07](https://github.com/user-attachments/assets/74b08c56-8fea-48e1-b66f-918f47200c4e)

I suspect this is because WinSCP is not being told to import the SSH Host CAs set in the PuTTY registry keys when it's invoked.

You can see under step 5 in the docs here that Teleport requires that checkbox enabled for seamless connections: https://goteleport.com/docs/connect-your-client/putty-winscp/#using-winscp-to-transfer-files-over-sftp

This is to save users from having to manually re-import their SSH Host CA configuration, which are also automatically generated and written to PuTTY's registry key when running `tsh puttyconfig`:

![image](https://github.com/user-attachments/assets/6935d582-73d8-4766-b19f-7dd588676f1d)

**Comment by WinSSHTerm** [comment]
Thanks for testing!

> This error is being passed back from Teleport, which does not allow X forwarding by default. As such, the PuTTY sessions it creates don't have X forwarding enabled when created either. It seems that out of the box, WinSSHTerm expects this to be enabled by default?

Right, X11 forwarding is enabled by default for all connections in WinSSHTerm by changing the session parameter in the registry. A way to get around this would be to adapt the migration script to turn off X11 forwarding for each migrated connection - if the session was created by Teleport. Maybe a simple substring check for "tsh.exe" in the proxy telnet command would be a good way to identify Teleport. What do you think?

> I suspect this is because WinSCP is not being told to import the SSH Host CAs set in the PuTTY registry keys when it's invoked.

That shouldn't be the case, because WinSSHTerm will force this by using the command line parameter `/rawconfig
Interface\SshHostCAsFromPuTTY=1`.

When you click on "Copy files" and the WinSCP windows pops up, you can see all command line parameters for WinSCP (and PuTTY) by opening up a PowerShell terminal and enter this command:
```
Get-CimInstance Win32_Process -Filter "name like '%putty%' or name like '%plink%' or name like '%winscp%'" | select CommandLine | Format-Table -wrap
```

With my test connection, this will ouput something like this for WinSCP:
```
"C:\Users\pst\...\tools\WinSCP\WinSCP.exe"
"scp:myuser@192.168.0.50:22/" /privatekey=C:\WinSSHTerm-2.37.0\my_user_private_key.ppk /rawconfig
Interface\SshHostCAsFromPuTTY=1 /rawsettings DetachedCertificate=C:\WinSSHTerm-2.37.0\my_user_certificate.pub
SendBuf=0 SshSimple=0 utf=1 /sessionname="srv01.mycompany.com (myuser@192.168.0.50)" /newinstance
```
The output might give a hint on what's going wrong here. Could you share it?

**Reply by webvictim** [reply]
>  Maybe a simple substring check for "tsh.exe" in the proxy telnet command would be a good way to identify Teleport. What do you think?

Yes, I think this would work well!

> WinSCP

Here's the command output (slightly redacted to change my Teleport server hostname):

```
"C:\Users\gus\Documents\WinSSHTerm\tools\WinSCP\WinSCP.exe" "scp:ubuntu@ip-172-31-30-140:3022/"
/privatekey=C:\Users\gus\.tsh\keys\teleport.example.com\webvictim.ppk /rawconfig Interface\SshHostCAsFromPuTTY=1
/rawsettings  ProxyMethod=5 ProxyHost=teleport.example.com ProxyPort=0 ProxyUsername=ubuntu ProxyPassword=
ProxyTelnetCommand="C:\\Windows\\tsh.exe proxy ssh --cluster=purple --proxy=%proxyhost %user@%host:%port" SendBuf=0
SshSimple=0 utf=1 /sessionname="ip-172-31-30-140%20(proxy:teleport.example.com) (ubuntu@ip-172-31-30-140)" /newinstance
```

**Reply by webvictim** [reply]
These are the "Raw Site Settings" from WinSCP for the equivalent PuTTY session which works, if it helps:

```
PublicKeyFile=C:%5CUsers%5Cgus%5C.tsh%5Ckeys%5Cteleport.example.com%5Cwebvictim.ppk
DetachedCertificate=C:%5CUsers%5Cgus%5C.tsh%5Ckeys%5Cteleport.example.com%5Cwebvictim-ssh%5Cpurple-cert.pub
LocalDirectory=C:%5C
RemoteDirectory=/home/ubuntu
ProxyMethod=5
ProxyHost=teleport.example.com
ProxyPort=0
ProxyUsername=ubuntu
ProxyTelnetCommand=C:\\Windows\\tsh.exe proxy ssh --cluster=purple --proxy=%proxyhost %user@%host:%port
```

Looks like `DetachedCertificate` is missing...

**Reply by WinSSHTerm** [reply]
Great, the output was very helpful. It seems the problem with WinSCP is caused by not passing the certificate as command line parameter. I made a test version 2.37.500 (arm64), where this bug should be fixed:
[WinSSHTerm-2.37.500.zip](https://github.com/user-attachments/files/16563449/WinSSHTerm-2.37.500.zip)

**Reply by webvictim** [reply]
This works better - the connection isn't rejected and the auth seems to work!

It now gets stuck here though:

![Screenshot 2024-08-09 at 12 04 45](https://github.com/user-attachments/assets/65bae1ee-7769-48f1-b498-888ea18c8e02)

After waiting a short while this comes up:

![Screenshot 2024-08-09 at 12 05 40](https://github.com/user-attachments/assets/005a58fa-ffcf-4c48-82f3-8932cfec8ed9)

I seem to recall during my original WinSCP testing that this happened when using SCP rather than SFTP. I can confirm that if I override "Protocol" to "sftp" under the "Copy Files" configuration - this resolves it and WinSCP connects as expected:

![image](https://github.com/user-attachments/assets/4c8dfcfc-9ef5-4378-907e-f81be5bcd8a5)

![Screenshot 2024-08-09 at 12 09 25](https://github.com/user-attachments/assets/4e62362a-40d6-477e-96b5-10c3163e2521)

**Reply by WinSSHTerm** [reply]
Thanks for testing! Nice, I will release an official fix soon.

> I seem to recall during my original WinSCP testing that this happened when using SCP rather than SFTP. I can confirm that if I override "Protocol" to "sftp" under the "Copy Files" configuration - this resolves it and WinSCP connects as expected

I'm not sure if I can do much here, as this doesn't seem to be a problem with WinSSHTerm. However, I could change the migration script, so that sftp will be set automatically for all migrated Teleport sessions. So no X11 Forwarding and SFTP instead of SCP. This should make connecting work out of the box. I will do some changes to the migration script and let you know.

**Reply by webvictim** [reply]
I think that sounds like a great solution!

**Reply by WinSSHTerm** [reply]
I made new releases:
[WinSSHTerm-2.37.2](https://github.com/WinSSHTerm/WinSSHTerm/releases/tag/2.37.2)
[Migrate2WinSSHTerm-0.20](https://github.com/P-St/Migrate2WinSSHTerm/releases/tag/0.20)
Connecting to PuTTY/WinSCP should work now without the need to change the configuration after migrating the sessions. Does it work for you?

**Reply by webvictim** [reply, marked as answer]
Hi - I can confirm that with WinSSHTerm 2.37.2 and Migrate2WinSSHTerm 0.20, everything works fine out of the box!
